src/gax/src/client_builder.rs

```diff
@@ -0,0 +1,678 @@
+// Copyright 2025 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Provide types for client construction.
+//!
+//! Some applications need to construct clients with custom configuration, for
+//! example, they may need to override the endpoint or the authentication
+//! credentials. The types in this module implement the client builders.
+//!
+//! Applications should not create builders directly, instead each client type
+//! defines a `builder()` function to obtain the correct type of builder:
+//!
+//! ## Example: create a client with the default configuration.
+//!
+//! ```
+//! # use google_cloud_gax::client_builder::examples;
+//! pub use examples::Client; // Placeholder for examples
+//! let client = Client::builder().build();
+//! ```
+//!
+//! ## Example: create a client with a different endpoint
+//!
+//! ```
+//! # use google_cloud_gax::client_builder::examples;
+//! pub use examples::Client; // Placeholder for examples
+//! let client = Client::builder()
+//!     .with_endpoint("https://private.googleapis.com")
+//!     .build();
+//! ```
+
+use crate::backoff_policy::{BackoffPolicy, BackoffPolicyArg};
+use crate::polling_backoff_policy::{PollingBackoffPolicy, PollingBackoffPolicyArg};
+use crate::polling_error_policy::{PollingErrorPolicy, PollingErrorPolicyArg};
+use crate::retry_policy::{RetryPolicy, RetryPolicyArg};
+use crate::retry_throttler::{RetryThrottlerArg, SharedRetryThrottler};
+use std::sync::Arc;
+
+#[derive(Clone, Debug)]
+pub struct ClientBuilder<F, Cr> {
+    config: internal::ClientConfig<Cr>,
+    factory: F,
+}
+
+impl<F, Cr> ClientBuilder<F, Cr> {
+    /// Creates a new client.
+    ///
+    /// ```
+    /// # use google_cloud_gax::client_builder::examples;
+    /// use examples::Client; // Placeholder for examples
+    /// let client = Client::builder()
+    ///     .build();
+    /// ```
+    pub fn build<C>(self) -> C
+    where
+        F: Fn(internal::ClientConfig<Cr>) -> C,
```

Comment:
Does this need to be `AsyncFn` ? and all the docs `.await`?

`Client::builder().build()` will call `async fn make_access_token_credential()` at some point.

Reply:
You are absolutely right.  I thought I could just change the function trait later, but it turns out fixing it changes everything else.  This is all solved now.